glue/src/main/scala/com/gu/tableversions/glue/GlueMetastore.scala

```diff
@@ -0,0 +1,184 @@
+      Option(glueTable.getPartitionKeys).toList.flatten map { glueColumn =>
+        PartitionColumn(glueColumn.getName)
+      }
+
+    def snapshotTableVersion(tableLocation: URI): TableVersion =
+      SnapshotTableVersion(VersionPaths.parseVersion(tableLocation))
+
+    for {
+      glueTable <- getGlueTable(table)
+      partitionColumns <- F.delay(getPartitionColumns(glueTable))
+      tableLocation <- F.delay(findTableLocation(glueTable))
+      version <- if (partitionColumns.isEmpty)
+        F.delay(snapshotTableVersion(tableLocation))
+      else
+        getPartitionedTableVersion(table, tableLocation, partitionColumns)
+    } yield (version)
+  }
+
+  private def getPartitionedTableVersion(
+      table: TableName,
+      tableLocation: URI,
+      partitionColumns: List[PartitionColumn]): F[TableVersion] = {
+
+    def toPartition(columnValues: List[ColumnValue]): Partition = columnValues match {
+      case head :: tail => Partition(head, tail: _*)
+      case _            => throw new Exception("empty columnValues list for partition found")
+    }
+
+    def toPartitionWithVersion(gluePartition: GluePartition): (Partition, Version) = {
+
+      val partitionColumnAndValue: List[(PartitionColumn, String)] = partitionColumns.zip(gluePartition.getValues)
+      val columnValues: List[ColumnValue] = partitionColumnAndValue.map { case (key, value) => ColumnValue(key, value) }
+      val partition: Partition = toPartition(columnValues)
+      val location = new URI(gluePartition.getStorageDescriptor.getLocation)
+      partition -> VersionPaths.parseVersion(location)
+    }
+
+    getGluePartitions(table).map { gluePartitions =>
+      val partitionVersions: Map[Partition, Version] = gluePartitions.map(toPartitionWithVersion).toMap
+      PartitionedTableVersion(partitionVersions)
+    }
+
+  }
+
+  override def update(table: TableName, changes: Metastore.TableChanges): F[Unit] =
+    changes.operations.traverse_(appliedOp(table))
+
+  private def appliedOp(table: TableName)(operation: TableOperation): F[Unit] =
+    operation match {
+      case AddPartition(partition, version)           => addPartition(table, partition, version)
+      case UpdatePartitionVersion(partition, version) => updatePartitionVersion(table, partition, version)
+      case RemovePartition(partition)                 => removePartition(table, partition)
+      case UpdateTableVersion(versionNumber)          => updateTableLocation(table, versionNumber)
+    }
+
+  def addPartition(table: TableName, partition: Partition, version: Version): F[Unit] = {
+
+    def partitionLocation(tableLocation: URI): String = {
+      val unversionedLocation: String = partition.resolvePath(tableLocation).toString
+      if (version == Version.Unversioned) unversionedLocation
+      else
+        unversionedLocation + version.label
+    }
+    for {
+      glueTable <- getGlueTable(table)
+      tableLocation <- F.delay(findTableLocation(glueTable))
```

Comment:
If `findTableLocation` is a pure function, you could do this instead here:

```scala
tableLocation = F.pure(findTableLocation(glueTable))
```

Or, if it's not pure (i.e. if it can throw exceptions) maybe it's worth moving the `F.delay` to the implementation of `findTableLocation`?

Reply:
I could just do `tableLocation = findTableLocation` if I'm not worried about exceptions right? 
I don't remember in this particular case but In most cases I added delay when I'm calling the aws java code that is out of our control so It might throw exceptions

Reply:
Yes I think it's very good to wrap calls to Java code in an effect. I just wonder if it would be worth doing it in the return type of `findTableLocation` instead of here, to do it as close to the Java code as possible, and maybe helping to ensure that one doesn't accidentally forget to do it in the caller. Minor point obviously.